modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/FilterNode.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.calcite.exec.rel;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.function.Predicate;
+import org.apache.ignite.internal.processors.query.calcite.exec.ExecutionContext;
+import org.apache.ignite.internal.util.typedef.F;
+
+/**
+ *
+ */
+public class FilterNode extends AbstractNode<Object[]> implements SingleNode<Object[]>, Upstream<Object[]> {
+    /** */
+    private final Predicate<Object[]> predicate;
+
+    /** */
+    private Deque<Object[]> inBuffer = new ArrayDeque<>(IN_BUFFER_SIZE);
```

Comment:
Why do we need buffer here and in other nodes? I agree that we need some kind of buffering in (or before) Outbox node, but anyway, I would do it as a separate tree node, like `BufferNode` for simplicity and flexibility.

Reply:
A filter node requests more rows than its consumer requests from the filter because the filter filters some of them) a small buffer is used to hold the difference till a next portion of rows is requested. The same is true for other nodes, where you may request from a node input more rows then requested from the node itself.